### PR TITLE
Fix various memory bugs detected by the address sanitizer

### PR DESCRIPTION
**Pull request by martyngigg**
**Description of work.**

Fixes a collection of memory errors detected by the address sanitizer in gcc 7.

**To test:**

Code review.

*There is no associated issue.*

*This does not require release notes* because **they are non-specific changes.**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Parallel/inc/MantidParallel/IO/EventParser.h [hunk 1]
```diff
@@ -219,7 +219,7 @@ void EventParser<TimeOffsetType>::startAsync(
   // Wrapped in lambda because std::thread is unable to specialize doParsing on
   // its own
   m_thread =
-      std::thread([this, event_id_start, event_time_offset_start, &range] {
+      std::thread([this, event_id_start, event_time_offset_start, range] {
```
Comment (martyngigg) [inline Framework/Parallel/inc/MantidParallel/IO/EventParser.h]: @SimonHeybrock Could I get your opinion on this change? Address sanitizer picked up a `use-after-scope` error in
[EventParserTest.h](https://github.com/mantidproject/mantid/blob/master/Framework/Parallel/test/EventParserTest.h#L229). `range` goes out of scope as the lambda survives longer than `startAsync` but range was a temporary so the reference becomes invalid.

As the range struct is small copying it into the lambda seemed the safest way to solve the issue as I don't think `startAsync` is performance critical?

Comment (SimonHeybrock) [reply Framework/Parallel/inc/MantidParallel/IO/EventParser.h]: I agree with your analysis, copying 24 Bytes is fine. :+1:

File: Framework/DataObjects/inc/MantidDataObjects/MDEventWorkspace.tcc [hunk 2]
```diff
@@ -54,9 +54,8 @@ Kernel::Logger logger("MDEventWorkspace");
 TMDE(MDEventWorkspace)::MDEventWorkspace(
     Mantid::API::MDNormalization preferredNormalization,
     Mantid::API::MDNormalization preferredNormalizationHisto)
-    : API::IMDEventWorkspace(), data(nullptr),
-      m_BoxController(new API::BoxController(nd)),
-      m_displayNormalization(preferredNormalization),
+    : API::IMDEventWorkspace(), m_BoxController(new API::BoxController(nd)),
+      data(nullptr), m_displayNormalization(preferredNormalization),
```
Comment (SimonHeybrock) [inline Framework/DataObjects/inc/MantidDataObjects/MDEventWorkspace.tcc]: Note that `unique_ptr` is `nullptr` by default, so no need for `data(nullptr)`?

File: Framework/DataObjects/inc/MantidDataObjects/MDEventWorkspace.tcc [hunk 3]
```diff
@@ -67,9 +66,8 @@ TMDE(MDEventWorkspace)::MDEventWorkspace(
 /** Copy constructor
  */
 TMDE(MDEventWorkspace)::MDEventWorkspace(const MDEventWorkspace<MDE, nd> &other)
-    : IMDEventWorkspace(other), data(nullptr),
-      m_BoxController(other.m_BoxController->clone()),
-      m_displayNormalization(other.m_displayNormalization),
+    : IMDEventWorkspace(other), m_BoxController(other.m_BoxController->clone()),
+      data(nullptr), m_displayNormalization(other.m_displayNormalization),
```
Comment (SimonHeybrock) [inline Framework/DataObjects/inc/MantidDataObjects/MDEventWorkspace.tcc]: dito.

File: qt/widgets/sliceviewer/src/LineViewer.cpp [hunk 4]
```diff
@@ -246,15 +246,15 @@ void LineViewer::updateFreeDimensions() {
...
+      m_startText[d]->setText(QString::number(m_start[d]));
+    if (d < m_end.getNumDims())
+      m_endText[d]->setText(QString::number(m_end[d]));
+    if (d < m_thickness.getNumDims())
+      m_thicknessText[d]->setText(QString::number(m_thickness[d]));
   }
-  ui.textPlaneWidth->setText(QString::number(m_planeWidth * 2.));
-
-  // Now show the width
-  this->updateBinWidth();
```
Comment (SimonHeybrock) [inline qt/widgets/sliceviewer/src/LineViewer.cpp]: I cannot see why this disappeared, can you elaborate?

**Review by SimonHeybrock — CHANGES_REQUESTED** [review]

File: Framework/DataObjects/inc/MantidDataObjects/MDEventWorkspace.tcc (hunk 2 above)
Comment (martyngigg) [reply Framework/DataObjects/inc/MantidDataObjects/MDEventWorkspace.tcc]: Good point. I copied without spotting that.

File: qt/widgets/sliceviewer/src/LineViewer.cpp (hunk 4 above)
Comment (martyngigg) [reply qt/widgets/sliceviewer/src/LineViewer.cpp]: You're right this should not have disappeared. I pushed some changes over the weekend that I hadn't tested yet. I will reinstate them.

**Comment by martyngigg** [comment]
@SimonHeybrock I think those concerns should be addressed now.

**Review by SimonHeybrock — APPROVED** [review]

**Pull request merged**